### PR TITLE
Add exclusive kwarg to scan ops

### PR DESCRIPTION
**Pull request by ibab**
This adds an `exclusive` kwarg to the `cumsum` and `cumprod` ops.
Documentation and tests have been extended to cover this kwarg.

### CODE REVIEW

File: tensorflow/core/ops/math_ops.cc [hunk 1]
```diff
@@ -1807,50 +1807,70 @@ product: Pairwise cross product of the vectors in `a` and `b`.
 REGISTER_OP("Cumsum")
     .Input("x: T")
     .Input("axis: int32")
+    .Attr("exclusive: bool")
```
Comment (vrv) [inline tensorflow/core/ops/math_ops.cc]: @girving, this op was just added, is it safe to add a new attr without a default?

Comment (girving) [inline tensorflow/core/ops/math_ops.cc]: It still needs a default since it's a separate PR.

File: tensorflow/python/ops/math_grad.py [hunk 2]
```diff
@@ -842,17 +842,19 @@ def _CrossGrad(op, grad):
 @ops.RegisterGradient("Cumsum")
 def _CumsumGrad(op, grad):
   axis = op.inputs[1]
+  exclusive = op.get_attr("exclusive")
   reverse = op.get_attr("reverse")
-  return [math_ops.cumsum(grad, axis=axis, reverse=(not reverse)), None]
+  return [math_ops.cumsum(grad, axis, exclusive, (not reverse)), None]
```
Comment (girving) [inline tensorflow/python/ops/math_grad.py]: Use keyword arg syntax for `exclusive` and `reverse`, and remove unnecessary parentheses around `not reverse`.

File: tensorflow/python/ops/math_grad.py [hunk 3]
```diff
@@ -842,17 +842,19 @@ def _CrossGrad(op, grad):
...
   x = op.inputs[0]
   axis = op.inputs[1]
+  exclusive = op.get_attr("exclusive")
   reverse = op.get_attr("reverse")
 
   # TODO This fails when x contains 0 and should be fixed
-  prod = math_ops.cumprod(x, axis=axis, reverse=reverse)
-  out = math_ops.cumsum(prod * grad, axis=axis, reverse=(not reverse))
+  prod = math_ops.cumprod(x, axis, exclusive, reverse)
+  out = math_ops.cumsum(prod * grad, axis, exclusive, (not reverse))
```
Comment (girving) [inline tensorflow/python/ops/math_grad.py]: `exclusive=exclusive`, `reverse=not reverse`

File: tensorflow/python/ops/math_ops.py [hunk 4]
```diff
@@ -1622,25 +1631,34 @@ def cumsum(x, axis=0, reverse=False, name=None):
     """
     with ops.op_scope([x], name, "Cumsum") as name:
       x = ops.convert_to_tensor(x, name="x")
-      return gen_math_ops.cumsum(x, axis, reverse, name=name)
+      return gen_math_ops.cumsum(x, axis, exclusive, reverse, name=name)
```
Comment (girving) [inline tensorflow/python/ops/math_ops.py]: Keyword arg syntax for exclusive and reverse.

File: tensorflow/python/ops/math_ops.py [hunk 5]
```diff
@@ -1656,7 +1674,7 @@ def cumprod(x, axis=0, reverse=False, name=None):
     """
     with ops.op_scope([x], name, "Cumprod") as name:
       x = ops.convert_to_tensor(x, name="x")
-      return gen_math_ops.cumprod(x, axis, reverse, name=name)
+      return gen_math_ops.cumprod(x, axis, exclusive, reverse, name=name)
```
Comment (girving) [inline tensorflow/python/ops/math_ops.py]: Keyword arg syntax for exclusive and reverse.

**Comment by girving** [comment]
Thank you for satisfying my inclusive scan pet peeve. :)

**Comment by ibab** [comment]
Okay, I've updated the PR.
I was trying to avoid line breaks by leaving out the kwargs, but I realize that it exactly a great solution :)

File: tensorflow/core/ops/math_ops.cc [hunk 6]
```diff
@@ -1842,7 +1842,7 @@ tf.cumsum([a, b, c], exclusive=True, reverse=True) ==> [b + c, c, 0]
 REGISTER_OP("Cumprod")
     .Input("x: T")
     .Input("axis: int32")
-    .Attr("exclusive: bool")
+    .Attr("exclusive: bool = false")
     .Attr("reverse: bool")
```
Comment (girving) [inline tensorflow/core/ops/math_ops.cc]: While we're at it, give reverse a false default too.

**Comment by ibab** [comment]
Okay, I've added a default for the reverse arg as well.

**Comment by girving** [comment]
Jenkins, test this please.

**Comment by girving** [comment]
I guess switch to keyword args for `_compareAll` to satisfy http://ci.tensorflow.org/job/tensorflow-pull-requests-sanity/771//console.

**Comment by ibab** [comment]
Yeah, I've done that but started to wonder why this didn't throw an error when running the tests.
Turns out my `for a, b, c in combinations([True, False], 3)` trick didn't actually do what I wanted and never executed the loop body :(
I've then noticed a few more problems with my tests after I corrected that.
I'll fix those and update the PR.

**Comment by girving** [comment]
Jenkins, test this please.

**Comment by ibab** [comment]
I've removed the 8D tests, as they were taking a very long time to run (past the timeout on my machine).

**Pull request merged**